gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,382 @@
+ *
+ * // Querying the result
+ * ObjectSet<GameObject> result = new ObjectSet<>();
+ * octree.query(cam.frustum, result);
+ *
+ * // Rendering the result
+ * for (GameObject gameObject : result) {
+ *   modelBatch.render(gameObject);
+ * }
+ * </pre>
+ */
+public class Octree<T> {
+
+	private final int maxItemsPerNode;
+
+	private final Pool<OctreeNode> nodePool = new Pool<OctreeNode>() {
+		@Override
+		protected OctreeNode newObject() {
+			return new OctreeNode();
+		}
+	};
+
+	protected OctreeNode root;
+	private final Collider<T> collider;
+
+	private static final Vector3 tmp = new Vector3();
+
+	public Octree (Vector3 minimum, Vector3 maximum, int maxDepth, int maxItemsPerNode, Collider<T> collider) {
+		super();
+		Vector3 realMin = new Vector3(Math.min(minimum.x, maximum.x), Math.min(minimum.y, maximum.y), Math.min(minimum.z, maximum.z));
+		Vector3 realMax = new Vector3(Math.max(minimum.x, maximum.x), Math.max(minimum.y, maximum.y), Math.max(minimum.z, maximum.z));
+
+		this.root = nodePool.obtain();
+		this.root.bounds.set(realMin, realMax);
+		this.root.level = maxDepth;
+		this.collider = collider;
+		this.maxItemsPerNode = maxItemsPerNode;
+	}
+
+	public void add (T object) {
+		root.add(object);
+	}
+
+	public void remove (T object) {
+		root.remove(object);
+	}
+
+	public void update (T object) {
+		root.remove(object);
+		root.add(object);
+	}
+
+	/** Method to retrieve all the geometries.
+	 * @param resultSet
+	 * @return the result set
+	 */
+	public ObjectSet<T> getAll (ObjectSet<T> resultSet) {
+		root.getAll(resultSet);
+		return resultSet;
+	}
+
+	/** Method to query geometries inside nodes that the aabb intersects.
+	 * Can be used as broad phase.
+	 * @param aabb - The bounding box to query
+	 * @param result - Set to be populated with objects inside the BoundingBoxes
+	 */
+	public ObjectSet<T> query (BoundingBox aabb, ObjectSet<T> result) {
+		root.query(aabb, result);
+		return result;
+	}
+
+	/** Method to query geometries inside nodes that the frustum intersects.
+	 * Can be used as broad phase.
+	 * @param frustum - The frustum to query
+	 * @param result set populated with objects near from the frustum
+	 */
+	public ObjectSet<T> query (Frustum frustum, ObjectSet<T> result) {
+		root.query(frustum, result);
+		return result;
+	}
+
+	public T rayCast (Ray ray, RayCastResult<T> result) {
+		result.distance = result.maxDistanceSq;
+		root.rayCast(ray, result);
+		return result.geometry;
+	}
+
+	/** Method to get nodes as bounding boxes. Useful for debug purpose.
+	 *
+	 * @param boxes
+	 */
+	public ObjectSet<BoundingBox> getNodesBoxes (ObjectSet<BoundingBox> boxes) {
+		root.getBoundingBox(boxes);
+		return boxes;
+	}
+
+	protected class OctreeNode {
+
+		private int level;
+
+		// Represents the boundary of the node
+		private final BoundingBox bounds;
+		private final Array<OctreeNode> children = new Array<>(8);
+		private final ObjectSet<T> geometries = new ObjectSet<T>(maxItemsPerNode);
```

Comment:
Note all nodes will allocate space to hold `maxItemsPerNode * 1.25` (as the default load factor is 0.8), even if they have much fewer or none. If it's not worth optimizing then just ignore this comment.

Does it need to be a set? I assume to avoid duplicates? If the number of items is low, an array may be sufficient. If a set is really needed to avoid n^2 when de-duping and the occtree is intended to be static, then maybe the set is only needed for construction. The drawback to using a set after construction is iteration (which appears to be done a lot) must check every entry in the backing table. One way to mitigate that is to use an OrderedSet, which keeps an array along with the set, then iteration uses the array. When using an OrderedSet for iteration when the order (ironically) doesn't actually matter, consider using `orderedItems().ordered = false;` for faster removal, as mentioned in the OrderedSet class javadocs.

Reply:
That's interesting, I can try to run a few (quick) benchmarks.

Thank you for the tip.

Reply:
@NathanSweet I changed to array. Thank you for the suggestion.

Reply:
problem is, Octree is not only used for static scene, we could have a lot of moving objects, not sure moving from ObjectSet to Array is a good idea.